### ISSUE TITLE
How to toggle line comments with mixed selection?

### ISSUE THREAD

**Issue by xeluxee** [numToStr/Comment.nvim#17]
Given a similar situation ↓ how to comment uncommented lines and uncomment commented ones?
``` c
printf("Line 1\n");
printf("Line 2\n");
// printf("Line 3\n");
// printf("Line 4\n");
```
Expected result: ↓
``` c
// printf("Line 1\n");
// printf("Line 2\n");
printf("Line 3\n");
printf("Line 4\n");
```
The result I get when I select these 4 lines and press `gc`: ↓
``` c
// printf("Line 1\n");
// printf("Line 2\n");
// // printf("Line 3\n");
// // printf("Line 4\n");
```

I think `gc` should toggle comments. To comment everything `g>` is already available...

**Comment by IndianBoy42** [comment]
btw this functionality can be added for visual mode very easily. using `:'<,'>normal gcc<CR>` (as a mapping use `:normal gcc<CR>`). This will run gcc on each line separately, toggling each lines comments. To use it as a operator you will need to do a bit more work dealing with `operatorfunc` but you can base its functionality on the visual mode mapping

**Comment by xeluxee** [comment]
Interesting, thanks
But how to use it in normal mode?
Something like `gt3j` to toggle comments on the next 3 lines?

**Comment by numToStr** [comment]
@xeluxee You might wanna use operatorfunc `:h operatorfunc`.

This might work though I haven't tested
```lua
A.nvim_set_keymap('n', 'gt', '<CMD>set operatorfunc=v:lua.__flip_flop_comment<CR>g@', { noremap = true, silent = true })
```

**Comment by VKondakoff** [comment]
> This is the best I can do to cover this case.

This does not seems to work here...
First I'm receiving a warning `Only has 3 variables, but you set 4 values` on this line:
```lua
local is_commented = U.is_commented(line, ll, rr, padding)
```

and then there is an error if I'm using `<leader>l` while there is visual selection: `E5108: Error executing lua ...ite/pack/packer/start/Comment.nvim/lua/Comment/utils.lua:259: attempt to concatenate local 'pp' (a boolean
 value)`

**Comment by numToStr** [comment]
So that now the API is stable. I am considering adding this but I am not sure whether to add a default mapping or just a give a lua api to play with.

**Comment by xeluxee** [comment]
Why not including it in extra mappings? A user can still disable it by setting `<nop>` if he don't like this feature

**Comment by numToStr** [comment]
What will be the keymap's LHS? I'll be similar to `gc` as it will also wait for some motion like `k`, `l`. I am also assuming that the user will also want this "flip-flop" to be used inside visual mode.

**Comment by xeluxee** [comment]
I thought to `gC` but maybe someone else could give us a better suggestion

> I am also assuming that the user will also want this "flip-flop" to be used inside visual mode.

Yeah it should be available in operator mode (`gC4j`) and in visual mode (`V4jgC`). I think it doesn't make sense to consider normal mode since to toggle comment on a single line `gcc` is already available.

**Comment by kuntau** [comment]
> What will be the keymap's LHS? I'll be similar to `gc` as it will also wait for some motion like `k`, `l`. I am also assuming that the user will also want this "flip-flop" to be used inside visual mode.

Yeah this. need it in operator pending mode. Visual mode I don't care much but still good to have 👍🏽

**Comment by hawkinchina** [comment]
Does it support  this form ?

printf("Line 1\n");
printf("Line 2\n");
printf("Line 3\n");
printf("Line 4\n");

=> 

/*
* printf("Line 1\n");
* printf("Line 2\n");
* printf("Line 3\n");
* printf("Line 4\n");
*/

**Comment by daangoossens22** [comment]
I updated the custom keymap described in https://github.com/numToStr/Comment.nvim/issues/17#issuecomment-939413608 since the internal util functions seem to have changed a little bit.
```lua
-- NOTE: custom workaround until natively supported
local U = require('Comment.utils')
function _G.__flip_flop_comment(opmode)
    local range = U.get_region(opmode)
    local lines = U.get_lines(range)
    local ctx = {
        ctype = U.ctype.line,
        range = range,
    }
    local cstr = require('Comment.ft').calculate(ctx) or vim.bo.commentstring
    local lcs, rcs = U.unwrap_cstr(cstr)
    local ll, rr = U.escape(lcs), U.escape(rcs)
    local padding, pp = U.get_padding(true)

    for i, line in ipairs(lines) do
        local is_commented = U.is_commented(ll, rr, pp)(line)
        if line == "" then
        elseif is_commented then
            lines[i] = U.uncomment_str(line, ll, rr, pp)
        else
            lines[i] = U.comment_str(line, lcs, rcs, padding)
        end
    end
    vim.api.nvim_buf_set_lines(0, range.srow - 1, range.erow, false, lines)
end
vim.keymap.set({ 'n', 'x' }, 'gC', '<cmd>set operatorfunc=v:lua.__flip_flop_comment<cr>g@', { silent = true, desc = "toggle the comment state of each line individually" })
```
It is not perfect but it does the job for me.
Although it would be nicer if this was provided as a extra keymap.

**Comment by xeluxee** [comment]
> It is not perfect but it does the job for me

Nice, thank you!
I've only found two issues using your function:
- Indentation: while commenting a block with different indentations comments are put before the first consistent character of every line. This leads formatters to align all these lines, breaking the original indentation. This doesn't happen with regular `gc`
- Visual selection: after selecting some text and pressing `gC` everything works fine, but selection marks are edited. So if I press `gv` only a smaller part of the previous selection gets selected. This doesn't happen with regular `gc`

**Comment by xeluxee** [comment]
> @xeluxee I quickly updated (by borrowing the indentation code from the actual plugin) it to address the issues you mentioned, hopefully that makes it also good enough for you as a bridging solution until its in the plugin itself (hopefully).

Now it works fine, thanks 😄

**Comment by bew** [comment]
Hello,
I was a user of NERDCommenter which used the terminology `invert` for this which I quite like, the mapping was ending with `i`, to show it's different than the `toggle` action.

For us, this could map to `gci` ?

**Comment by numToStr** [comment]
> For us, this could map to `gci`?

No, because it would conflict with `gci{motion}` like `gciw`, `gcip` etc.

**Comment by bew** [comment]
Oh right ><
`gi` ? (but that's also a useful builtin..)

What about no builtin mapping and let people set one if needed? 

Is there a blocker for this to end up in the plugin ? (except finding a mapping for it :eyes:)

**Comment by numToStr** [comment]
> What about no builtin mapping and let people set one if needed?

That's also my thought to just expose the API.

> Is there a blocker for this to end up in the plugin ?

$DAYJOB and time :)

**Comment by hiberabyss** [comment]
Temporary solution with vmap:

```
vmap <silent> gC :normal gcc<cr>
```

**Comment by mmirus** [comment]
@daangoossens22 Thanks for sharing your solution! Unfortunately, I get an error when I try it. Any chance you have an updated version you could share, if you also hit this?

```
E5108: Error executing lua ...ite/pack/packer/start/Comment.nvim/lua/Comment/utils.lua:149: bad argument #1 to 'match' (string expected, got table)
stack traceback:
        [C]: in function 'match'
        ...ite/pack/packer/start/Comment.nvim/lua/Comment/utils.lua:149: in function 'unwrap_cstr'
```

**Comment by xeluxee** [comment]
@numToStr is there a way to determine if a line is commented using new APIs?
Unfortunately @daangoossens22's solution uses internal functions that are likely to be changed, leading to unexpected errors

**Comment by numToStr** [comment]
@xeluxee You can use `:h comment.utils.is_commented`

```lua
require'Comment.utils'.is_commented('--', '', true)('   -- Hello')
```

(I hope to start working on this after neovim 0.8 release which is just couple of days away)

**Comment by mmirus** [comment]
TY @xeluxee! :partying_face:

**Comment by OliverChao** [comment]
Thanks a lot @xeluxee. I want that function as well. :clap:
